src/Stratis.Bitcoin.Features.Consensus.Tests/TimeSyncBehaviorAdjustedTimeTest.cs

```diff
@@ -0,0 +1,96 @@
+        
+        /// <summary>
+        /// This is the result of the operation under test. 
+        /// </summary>
+        private TimeSpan offset;
+
+        /// <summary>
+        /// Checks that <see cref="TimeSyncBehaviorState.AddTimeData(IPAddress, TimeSpan, bool)"/>
+        /// uses outbound samples as a priority over inbound as they are less likely to be malicious.
+        /// This means that even when there are many inbound connections, only the first ones are considered until there are enough outbound nodes. 
+        /// Eg only 9 inbound per 1 outbound.
+        /// </summary>
+        [Fact]
+        public void AddTimeData_WithLargeSampleSetOfInboundTimeManipulatorsAndLowSampleSetOfOutbound_GetsOverridenByOutboundSamples()
+        {
+            this.Given_an_empty_time_sync_behaviour_state();
+            this.Given_40_inbound_samples_with_offset_of(10);
+            this.Given_1_outbound_sample_with_offset_of(20);
+            this.When_calculating_time_adjust_offset();
+            this.Then_adjusted_time_offset_should_be(20);
+        }
+
+        /// <summary>
+        /// Checks that <see cref="TimeSyncBehaviorState.AddTimeData(IPAddress, TimeSpan, bool)"/>
+        /// uses system time without an offset when there are no outbound samples.
+        /// </summary>
+        [Fact]
+        public void AddTimeData_WithLargeSampleSetOfInboundTimeManipulatorsAndZeroOutbound_SticksWithTheSystemTime()
+        {
+            this.Given_an_empty_time_sync_behaviour_state();
+            this.Given_40_inbound_samples_with_offset_of(10);
+            this.When_calculating_time_adjust_offset();
+            this.Then_adjusted_time_offset_should_be(0);
+        }
+
+        private void Given_an_empty_time_sync_behaviour_state()
+        {
+            this.dateTimeProvider = new DateTimeProvider();
+            this.timesyncBehaviourState = new TimeSyncBehaviorState(
+                this.dateTimeProvider, new NodeLifetime(),
+                new AsyncLoopFactory(new LoggerFactory()), 
+                new LoggerFactory());
+        }
+
+        private void Given_1_outbound_sample_with_offset_of(int offsetSeconds)
+        {
+            this.timesyncBehaviourState.AddTimeData(IPAddress.Parse("2.2.2.2"), TimeSpan.FromSeconds(offsetSeconds), isInboundConnection: false);
+        }
+
+        private void Given_40_inbound_samples_with_offset_of(int offset)
+        {
+            for (int i = 1; i <= 40; i++)
+            {
+                this.timesyncBehaviourState.AddTimeData(IPAddress.Parse("1.2.3." + i), TimeSpan.FromSeconds(offset), isInboundConnection: true);
+            }
+        }
+
+        private void When_calculating_time_adjust_offset()
+        {
+            var adjustedTime = this.dateTimeProvider.GetAdjustedTime();
+            var now = this.dateTimeProvider.GetUtcNow();
+            this.offset = adjustedTime - now;
+        }
+
+        private void Then_adjusted_time_offset_should_be(int expectedOffset)
```

Comment:
This method accepts the value which is used for accretion. That way you are kind of hiding what you are checking in the inner code. 
 
Why not instead create a method `getRoundedOffcet` and use 

`Assert.Equal(expectedOffset, getRoundedOffcet());` in placed where you are using `Then_adjusted_time_offset_should_be` now

That way it will be clear from a high level code what you are checking.

Reply:
I see what you are saying, but,  Given When Then is a pattern for tests. To do Given When Assert  breaks that pattern.....

I changed it to this way of having to understand the example parameter and looking at the overall method rather than this line in isolation.... it feels very clear to me with these parameters and knowing the answer - then the overall method name gives the why......

Reply:
Ok. 
I still would like to change that but let's wait until all problems are solved and there is only that pattern under the question.